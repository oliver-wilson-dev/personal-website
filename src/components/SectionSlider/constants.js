/*
  At the time of writing these two (X_AXIS and Y_AXIS) are the same,
  I am keeping them as stand alone variables to allow for easy customisation
  in the future if required.
*/
const X_AXIS = 75;
const Y_AXIS = 75;

export const PX_SWIPED_THRESHOLD = {
  X_AXIS,
  Y_AXIS
};
