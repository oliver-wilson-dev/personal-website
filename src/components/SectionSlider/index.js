/* eslint-disable react/no-array-index-key */
import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import SectionTile from '../SectionTile';
import styles from './index.css';
import sectionStyles from '../SectionTile/index.css';
import LeftArrow from './left-arrow.svg';
import RightArrow from './right-arrow.svg';


const SectionSlider = ({ children }) => {
  const [currentSelectedIndex, setCurrentSelectedIndex] = useState(0);
  const [touchStart, setTouchStart] = useState(null);
  const [touchEnd, setTouchEnd] = useState(null);

  const firstElementSelected = currentSelectedIndex === 0;

  const lastElementSelected = currentSelectedIndex === children.length - 1;

  const increment = useCallback(() => {
    if (!lastElementSelected) {
      setCurrentSelectedIndex(currentSelectedIndex + 1);
    }
    setTouchStart(null);
    setTouchEnd(null);
  }, [currentSelectedIndex, lastElementSelected]);

  const decrement = useCallback(() => {
    if (!firstElementSelected) {
      setCurrentSelectedIndex(currentSelectedIndex - 1);
    }
    setTouchStart(null);
    setTouchEnd(null);
  }, [currentSelectedIndex, firstElementSelected]);

  const onTouchStart = useCallback((e) => {
    e.stopPropagation();
    setTouchStart(e.targetTouches[0].clientX);
  }, []);

  const onTouchEnd = useCallback((e) => {
    e.stopPropagation();
    setTouchEnd(e.changedTouches[0].clientX);
  }, []);

  useEffect(() => {
    if (touchStart !== null && touchEnd !== null) {
      if (touchStart - touchEnd > 75) {
        increment();
      }

      if (touchStart - touchEnd < -75) {
        decrement();
      }
    }
  }, [touchStart, touchEnd, increment, decrement]);

  return (
    <React.Fragment>
      <div className={styles.buttonWrapper}>
        <LeftArrow
          className={cn(styles.arrowBtn, {
            [styles.btnDisabled]: firstElementSelected
          })}
          onClick={decrement}
        />
        <RightArrow
          className={cn(styles.arrowBtn, {
            [styles.btnDisabled]: lastElementSelected
          })}
          onClick={increment}
        />
      </div>
      <div
        className={cn(
          styles.tileContainer,
          sectionStyles.displayBox
        )}
        onTouchStart={onTouchStart}
        onTouchEnd={onTouchEnd}
      >
        { children.map((child, index) => {
          const props = {
            border: true,
            additionalStyles: {
              [styles.setWidth]: true,
              [styles.tileOnScreen]: index === currentSelectedIndex,
              [styles.offScreenTileRight]: index > currentSelectedIndex,
              [styles.offScreenTileLeft]: index < currentSelectedIndex
            }
          };

          return (
            <SectionTile key={index} {...props}>
              {child}
            </SectionTile>
          );
        })}
      </div>
    </React.Fragment>
  );
};

SectionSlider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export default SectionSlider;
