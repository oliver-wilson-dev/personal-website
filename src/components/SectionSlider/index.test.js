import React from 'react';
import { shallow, mount } from 'enzyme';
import SectionSlider from '.';
import styles from './index.css';

jest.mock('../SectionTile', () => {
  const SectionTile = () => <div />;

  return SectionTile;
});

jest.mock('./left-arrow.svg', () => {
  const LeftArrow = () => null;

  return LeftArrow;
});

jest.mock('./right-arrow.svg', () => {
  const RightArrow = () => null;

  return RightArrow;
});

const defaultChildren = [
  <p key="1">hello world</p>,
  <p key="2">goodbye world</p>
];

const render = (children = defaultChildren, renderMethod = shallow) => renderMethod(
  <SectionSlider>
    {children}
  </SectionSlider>
);

describe('SectionSlider', () => {
  it('should render correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  describe('when clicking the right arrow', () => {
    it('should apply the btnDisabled class to the right arrow', () => {
      const component = render();

      component.find(`.${styles.arrowBtn}`).at(1).simulate('click');

      component.update();

      expect(component.find(`.${styles.arrowBtn}`).at(1).hasClass(styles.btnDisabled)).toBe(true);
    });

    describe('and then clicking it again', () => {
      it('should render correctly', () => {
        const component = render();

        component.find(`.${styles.arrowBtn}`).at(1).simulate('click');

        component.update();

        component.find(`.${styles.arrowBtn}`).at(1).simulate('click');

        component.update();

        expect(component).toMatchSnapshot();
      });
    });

    describe('and then clicking the left arrow', () => {
      it('should apply the btnDisabled class to the left arrow', () => {
        const component = render();

        component.find(`.${styles.arrowBtn}`).at(1).simulate('click');

        component.update();

        component.find(`.${styles.arrowBtn}`).at(0).simulate('click');

        component.update();

        expect(component.find(`.${styles.arrowBtn}`).at(0).hasClass(styles.btnDisabled)).toBe(true);
      });

      describe('and then clicking it again', () => {
        it('should render correctly', () => {
          const component = render();

          component.find(`.${styles.arrowBtn}`).at(1).simulate('click');

          component.update();

          component.find(`.${styles.arrowBtn}`).at(0).simulate('click');

          component.update();

          component.find(`.${styles.arrowBtn}`).at(0).simulate('click');

          component.update();

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('when swiping left and then right', () => {
      it('should show correct tiles', () => {
        const component = render(defaultChildren, mount);

        const touchStart = (x = '0', y = '0') => {
          component.find(`.${styles.tileContainer}`).simulate('touchStart', { targetTouches: [{ clientX: x, clientY: y }], stopPropagation: jest.fn() });
          component.update();
        };

        const touchEnd = (x = '0', y = '0') => {
          component.find(`.${styles.tileContainer}`).simulate('touchEnd', { changedTouches: [{ clientX: x, clientY: y }], stopPropagation: jest.fn() });
          component.update();
        };

        touchStart('0');
        touchEnd('200');


        expect(component.find(`.${styles.arrowBtn}`).at(0).hasClass(styles.btnDisabled)).toBe(true);
        expect(component.find(`.${styles.arrowBtn}`).at(1).hasClass(styles.btnDisabled)).toBe(false);

        touchStart('200');
        touchEnd('0');

        expect(component.find(`.${styles.arrowBtn}`).at(0).hasClass(styles.btnDisabled)).toBe(false);
        expect(component.find(`.${styles.arrowBtn}`).at(1).hasClass(styles.btnDisabled)).toBe(true);
      });
    });
    describe('when swiping and the swipe is less than 75px', () => {
      it('should show correct tiles', () => {
        const component = render(defaultChildren, mount);

        const touchStart = (x = '0', y = '0') => {
          component.find(`.${styles.tileContainer}`).simulate('touchStart', { targetTouches: [{ clientX: x, clientY: y }], stopPropagation: jest.fn() });
          component.update();
        };

        const touchEnd = (x = '0', y = '0') => {
          component.find(`.${styles.tileContainer}`).simulate('touchEnd', { changedTouches: [{ clientX: x, clientY: y }], stopPropagation: jest.fn() });
          component.update();
        };

        touchStart('0');
        touchEnd('50');


        expect(component.find(`.${styles.arrowBtn}`).at(0).hasClass(styles.btnDisabled)).toBe(true);
        expect(component.find(`.${styles.arrowBtn}`).at(1).hasClass(styles.btnDisabled)).toBe(false);

        touchStart('50');
        touchEnd('-10');

        expect(component.find(`.${styles.arrowBtn}`).at(0).hasClass(styles.btnDisabled)).toBe(true);
        expect(component.find(`.${styles.arrowBtn}`).at(1).hasClass(styles.btnDisabled)).toBe(false);
      });
    });
  });
});
