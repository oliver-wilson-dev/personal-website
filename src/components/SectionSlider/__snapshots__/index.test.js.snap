// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionSlider should render correctly 1`] = `
<Fragment>
  <div
    className="buttonWrapper"
  >
    <LeftArrow
      className="arrowBtn btnDisabled"
      onClick={[Function]}
    />
    <RightArrow
      className="arrowBtn"
      onClick={[Function]}
    />
  </div>
  <div
    className="tileContainer displayBox"
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <SectionTile
      additionalStyles={
        Object {
          "offScreenTileLeft": false,
          "offScreenTileRight": false,
          "setWidth": true,
          "tileOnScreen": true,
        }
      }
      border={true}
      key="0"
    >
      <p
        key="1"
      >
        hello world
      </p>
    </SectionTile>
    <SectionTile
      additionalStyles={
        Object {
          "offScreenTileLeft": false,
          "offScreenTileRight": true,
          "setWidth": true,
          "tileOnScreen": false,
        }
      }
      border={true}
      key="1"
    >
      <p
        key="2"
      >
        goodbye world
      </p>
    </SectionTile>
  </div>
</Fragment>
`;

exports[`SectionSlider when clicking the right arrow and then clicking it again should render correctly 1`] = `
<Fragment>
  <div
    className="buttonWrapper"
  >
    <LeftArrow
      className="arrowBtn"
      onClick={[Function]}
    />
    <RightArrow
      className="arrowBtn btnDisabled"
      onClick={[Function]}
    />
  </div>
  <div
    className="tileContainer displayBox"
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <SectionTile
      additionalStyles={
        Object {
          "offScreenTileLeft": true,
          "offScreenTileRight": false,
          "setWidth": true,
          "tileOnScreen": false,
        }
      }
      border={true}
      key="0"
    >
      <p
        key="1"
      >
        hello world
      </p>
    </SectionTile>
    <SectionTile
      additionalStyles={
        Object {
          "offScreenTileLeft": false,
          "offScreenTileRight": false,
          "setWidth": true,
          "tileOnScreen": true,
        }
      }
      border={true}
      key="1"
    >
      <p
        key="2"
      >
        goodbye world
      </p>
    </SectionTile>
  </div>
</Fragment>
`;

exports[`SectionSlider when clicking the right arrow and then clicking the left arrow and then clicking it again should render correctly 1`] = `
<Fragment>
  <div
    className="buttonWrapper"
  >
    <LeftArrow
      className="arrowBtn btnDisabled"
      onClick={[Function]}
    />
    <RightArrow
      className="arrowBtn"
      onClick={[Function]}
    />
  </div>
  <div
    className="tileContainer displayBox"
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
  >
    <SectionTile
      additionalStyles={
        Object {
          "offScreenTileLeft": false,
          "offScreenTileRight": false,
          "setWidth": true,
          "tileOnScreen": true,
        }
      }
      border={true}
      key="0"
    >
      <p
        key="1"
      >
        hello world
      </p>
    </SectionTile>
    <SectionTile
      additionalStyles={
        Object {
          "offScreenTileLeft": false,
          "offScreenTileRight": true,
          "setWidth": true,
          "tileOnScreen": false,
        }
      }
      border={true}
      key="1"
    >
      <p
        key="2"
      >
        goodbye world
      </p>
    </SectionTile>
  </div>
</Fragment>
`;
